packages/legend-graph/src/models/protocols/pure/v1/V1_PureGraphManager.ts

```diff
@@ -2016,6 +2036,237 @@ export class V1_PureGraphManager extends AbstractPureGraphManager {
+    return graphBuilderInput;
+  }
+
+  async buildCreateQueryLightGraph(
+    graph: PureModel,
+    entities: Entity[],
+    dependencyEntities: Map<string, Entity[]>,
+  ): Promise<void> {
+    try {
+      const graphBuilderInput = await this.indexEntitiesWithDependencyIntoGraph(
+        graph,
+        entities,
+        dependencyEntities,
+        [V1_Mapping, V1_PackageableRuntime],
+        [
+          V1_MAPPING_ELEMENT_PROTOCOL_TYPE,
+          V1_PACKAGEABLE_RUNTIME_ELEMENT_PROTOCOL_TYPE,
+        ],
+      );
+      // handle mapping includes
+      const mappings = [
+        ...graph.ownMappings,
+        ...graph.dependencyManager.mappings,
+      ];
+      const v1Mappings = graphBuilderInput
+        .map((e) =>
+          e.data.elements.filter(
+            (type: V1_PackageableElement): type is V1_Mapping =>
+              type instanceof V1_Mapping,
+          ),
+        )
+        .flat();
+      const context = new V1_GraphBuilderContextBuilder(
+        graph,
+        graph,
+        this.extensions,
+        this.log,
+      ).build();
+      // build include index for compatible runtime analysis
+      v1Mappings.forEach((element) => {
+        const mapping = mappings.find((e) => e.path === element.path);
+        if (mapping) {
+          mapping.includes = element.includedMappings.map((i) => {
+            assertNonEmptyString(
+              i.includedMappingPath,
+              `Mapping include 'includedMappingPath' field is missing or empty`,
+            );
+            return V1_buildMappingInclude(i, context, mapping);
+          });
+        }
+      });
+      // handle runtimes
+      const runtimes = [
+        ...graph.ownRuntimes,
+        ...graph.dependencyManager.runtimes,
+      ];
+      const v1Runtimes = graphBuilderInput
+        .map((e) =>
+          e.data.elements.filter(
+            (type: V1_PackageableElement): type is V1_PackageableRuntime =>
+              type instanceof V1_PackageableRuntime,
+          ),
+        )
+        .flat();
+      v1Runtimes.forEach((element) => {
+        const runtime = runtimes.find((e) => e.path === element.path);
+        if (runtime) {
+          const runtimeValue = new EngineRuntime();
+          runtime.setRuntimeValue(runtimeValue);
+          runtimeValue.setMappings(
+            element.runtimeValue.mappings.map((mapping) =>
+              context.resolveMapping(mapping.path),
+            ),
+          );
+        }
+      });
+      graph.buildState.pass();
+    } catch (error) {
+      assertErrorThrown(error);
+      graph.buildState.fail();
+      /**
+       * Wrap all error with `GraphBuilderError`, as we throw a lot of assertion error in the graph builder
+       * But we might want to rethink this decision in the future and throw appropriate type of error
+       */
+      throw error instanceof GraphBuilderError
+        ? error
+        : new GraphBuilderError(error);
+    }
+  }
+
+  async buildServiceQueryLightGraph(
```

Comment:
Could we put a comment to keep my suggestion about optimizing this further, i.e. to tinker with the entities to omit parts we don't need? Thanks!